### PR TITLE
[CALCITE-2317] Support JDBC DatabaseMetaData.getFunctions

### PR DESCRIPTION
**Pull request by MalteBellmann**
DatabaseMetaData.getFunctions() on a connection to Calcite now returns UDFs.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/jdbc/CalciteMetaImpl.java [hunk 1]
```diff
@@ -492,6 +493,48 @@ public MetaResultSet getTableTypes(ConnectionHandle ch) {
...
+                        (short) DatabaseMetaData.functionResultUnknown,
+                        name
+                    )
+                )
+        );
+  }
+
+  Enumerable<MetaFunction> functions(final MetaSchema schema, final String catalog,
+      final Predicate1<String> functionNameMatcher) {
+    return functions(schema, catalog)
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/jdbc/CalciteMetaImpl.java:534]: Is it possible to add some tests?

Comment (MalteBellmann) [reply core/src/main/java/org/apache/calcite/jdbc/CalciteMetaImpl.java:538]: Thanks for reviewing, i've added a test and sorting according to the JDBC specification. 

FYI, there is also a discussion in JIRA on how to handle built-in functions.

**Comment by MalteBellmann** [comment]
@chunweilei do you have time to review this updated PR? I think it is ready to merge

File: core/src/main/java/org/apache/calcite/jdbc/CalciteMetaImpl.java [hunk 2]
```diff
@@ -502,6 +507,70 @@ CalciteConnectionImpl getConnection() {
...
+        "FUNCTION_NAME",
+        "REMARKS",
+        "FUNCTION_TYPE",
+        "SPECIFIC_NAME");
+  }
+
+  Enumerable<MetaFunction> functions(final MetaSchema schema_, final String catalog) {
+    final CalciteMetaSchema schema = (CalciteMetaSchema) schema_;
+    Enumerable<MetaFunction> opTableFunctions = Linq4j.emptyEnumerable();
+    if (schema.getName().equals("metadata")) {
```
Comment (amaliujia) [inline core/src/main/java/org/apache/calcite/jdbc/CalciteMetaImpl.java:534]: Out of curiosity: why hardcode `metadata` as schema name?

**Comment by amaliujia** [comment]
Overall makes sense to me

File: core/src/main/java/org/apache/calcite/jdbc/CalciteMetaImpl.java (hunk 2 above)
Comment (MalteBellmann) [reply core/src/main/java/org/apache/calcite/jdbc/CalciteMetaImpl.java:534]: The idea was: It is hardcoded in https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/jdbc/CalciteSchema.java#L525-L527 and I didn't see a better way.

But I just realized, I can use `MetadataSchema.INSTANCE` for the check. I will update the PR accordingly

**Review by amaliujia — APPROVED** [review]

File: core/src/test/java/org/apache/calcite/test/JdbcAdapterTest.java [hunk 3]
```diff
@@ -760,6 +761,47 @@
...
+        + "         {\n"
+        + "           table: 'TEST_VIEW',\n"
+        + "           sql: 'SELECT 1'\n"
+        + "         }\n"
+        + "       ]\n"
+        + "     }\n"
+        + "   ]\n"
+        + "}";
+    CalciteAssert.model(model)
+        .withDefaultSchema("adhoc")
```
Comment (xy2953396112) [inline core/src/test/java/org/apache/calcite/test/JdbcAdapterTest.java:792]: LGTM

**Comment by amaliujia** [comment]
Merged via https://github.com/apache/calcite/commit/e447ff85a0a7d2037b819b7ced91177261974c7e

**Pull request closed without merging**